### PR TITLE
Refactor job queues - Closes #221

### LINKED ISSUE #221: Refactor job queues
### Description

Prepare universal implementation of Redis-based job queues. The transaction statistics are build based on queues, however that code is not universal and cannot be reused. Make sure the queues can be used in other scenarios.

### Motivation

- Job queues can be leveraged in many other scenarios such as building search index

### Acceptance Criteria

- Job queue can be initialised
- A task can be added to the queue

### PR DESCRIPTION
**Pull request by nagdahimanshu**
### What was the problem?
This PR resolves #221 

### How was it solved?
- [x] Refactor job queues to make generic
- [x] Job queue can be initialised
- [x] A task can be added to the queue

### How was it tested?
Local + Jenkins

### CODE REVIEW

File: services/core/shared/core/queue.js [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+const { Logger } = require('lisk-service-framework');
+const Queue = require('bull');
+const util = require('util');
+
+const config = require('../../config');
+
+const logger = Logger();
+
+const initializeQueue = (queueName, queueProcess) => {
+	const statsQueue = new Queue(queueName, {
```
Comment (MichalTuleja) [inline services/core/shared/core/queue.js:25]: Could you please make that variable name more generic?

File: services/core/shared/core/queue.js [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+		redis: config.endpoints.redis,
+		limiter: {
+			max: 8,
+			duration: 20,
+		},
+		prefix: 'Queue',
```
Comment (MichalTuleja) [inline services/core/shared/core/queue.js:31]: The queue prefix could be taken from `package.json -> name`.
Like `queue-${packageJsonName}` or so.

**Review by MichalTuleja — CHANGES_REQUESTED** [review]
That implementation looks great 👍 
I've left two minor remarks.

File: services/core/shared/core/queue.js (hunk 1 above)
Comment (nagdahimanshu) [reply services/core/shared/core/queue.js]: I made it generic by making given values configurable based on different queues.

File: services/core/shared/core/queue.js (hunk 2 above)
Comment (nagdahimanshu) [reply services/core/shared/core/queue.js]: Done.

File: services/core/config.js [hunk 3]
```diff
@@ -149,4 +149,31 @@ config.log.file = process.env.SERVICE_LOG_FILE || 'false';
 config.log.docker_host = process.env.DOCKER_HOST || 'local';
 config.debug = process.env.SERVICE_LOG_LEVEL === 'debug';
 
+config.queue = {
+	defaults: {
+		limiter: {
+			max: 8,
+			duration: 20,
```
Comment (sameersubudhi) [inline services/core/config.js:156]: A comment would be appreciated to indicate the default time units.

```suggestion
			duration: 20, // millisecs
```
or
```suggestion
			duration: 20, // secs
```

File: services/core/config.js [hunk 4]
```diff
@@ -149,4 +149,31 @@ config.log.file = process.env.SERVICE_LOG_FILE || 'false';
...
+		},
+		defaultJobOptions: {
+			attempts: 5,
+			timeout: 5 * 60 * 1000,
```
Comment (sameersubudhi) [inline services/core/config.js:160]: Same as above!
```suggestion
			timeout: 5 * 60 * 1000,  // millisecs
```

File: services/core/shared/core/queue.js [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+const Queue = require('bull');
+const util = require('util');
+
+const packageJson = require('../../package.json');
+const config = require('../../config');
+
+const logger = Logger();
+
+const initializeQueue = (queueName, queueProcess, options) => {
+	const queue = new Queue(queueName, {
```
Comment (sameersubudhi) [inline services/core/shared/core/queue.js:26]:
```suggestion
const initializeQueue = (queueName = 'defaultQueue', queueProcess, options = config.queue.defaults) => {
	queueProcess = queueProcess ? queueProcess : () => {}; // Please test this
	
	const queue = new Queue(queueName, {
```

File: services/core/shared/core/transactionStatistics.js [hunk 6]
```diff
@@ -14,11 +14,17 @@
...
 
 const config = require('../../config');
-const transactionStatisticsQueue = require('./transactionStatisticsQueue');
+const requestAll = require('../requestAll');
+const { getTransactions } = require('./transactions');
+const { initializeQueue } = require('./queue');
 const getDbInstance = require('../pouchdb');
 
+const queueName = 'transactionStatisticsQueue';
+const queueOptions = config.queue[queueName] || config.queue.defaults;
```
Comment (sameersubudhi) [inline services/core/shared/core/transactionStatistics.js:26]: Default values should automatically be picked up if specific configs are not defined. Please check the suggested changes in `services/core/shared/core/queue.js`.
```suggestion
const queueOptions = config.queue[queueName];
```

File: services/core/shared/core/queue.js [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+		redis: config.endpoints.redis,
+		limiter: options.limiter,
+		prefix: `queue-${packageJson.name}`,
+		defaultJobOptions: options.defaultJobOptions,
+		settings: options.settings,
+	});
+
+	queue.process(queueName, queueProcess);
```
Comment (sameersubudhi) [inline services/core/shared/core/queue.js:34]:
```suggestion
	queue.process(queueName, queueJob);
```

**Review by sameersubudhi — CHANGES_REQUESTED** [review]
Kindly check my suggestions.

File: services/core/config.js (hunk 3 above)
Comment (nagdahimanshu) [reply services/core/config.js]: Done!

File: services/core/config.js (hunk 4 above)
Comment (nagdahimanshu) [reply services/core/config.js]: Done!

File: services/core/shared/core/queue.js (hunk 5 above)
Comment (nagdahimanshu) [reply services/core/shared/core/queue.js]: Code updated to:

```
const initializeQueue = (queueName = 'defaultQueue', queueJob, options = config.queue.defaults) => {
	const queue = new Queue(queueName, {
```

File: services/core/shared/core/queue.js (hunk 7 above)
Comment (nagdahimanshu) [reply services/core/shared/core/queue.js]: Done!

File: services/core/shared/core/transactionStatistics.js (hunk 6 above)
Comment (nagdahimanshu) [reply services/core/shared/core/transactionStatistics.js]: Done!

**Review by sameersubudhi — APPROVED** [review]
Thank you! Great job 👏

**Review by MichalTuleja — APPROVED** [review]
LGTM 👍

**Pull request merged**